crypto/asn1/i2d_evp.c

```diff
@@ -0,0 +1,146 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* We need to use some deprecated APIs to support the legacy bits */
+#define OPENSSL_SUPPRESS_DEPRECATED
+
+#include <stdio.h>
+#include "internal/cryptlib.h"
+#include <openssl/evp.h>
+#include <openssl/encoder.h>
+#include <openssl/buffer.h>
+#include <openssl/x509.h>
+#include <openssl/rsa.h>         /* For i2d_RSAPublicKey */
+#include <openssl/dsa.h>         /* For i2d_DSAPublicKey */
+#include <openssl/ec.h>          /* For i2o_ECPublicKey */
+#include "crypto/asn1.h"
+#include "crypto/evp.h"
+
+static int i2d_provided(OSSL_LIB_CTX *libctx,
+                        const EVP_PKEY *a, const char *propq,
+                        int selection, const char *output_structures[],
+                        unsigned char **pp)
+{
+    OSSL_ENCODER_CTX *ctx = NULL;
+    int ret;
+
+    for (ret = -1;
+         ret == -1 && *output_structures != NULL;
+         output_structures++) {
+        /*
+         * The i2d_ calls don't take a boundary length for *pp.  However,
+         * OSSL_ENCODER_CTX_get_num_encoders() needs one, so we make one
+         * up.
+         */
+        size_t len = INT_MAX;
+
+        ctx = OSSL_ENCODER_CTX_new_by_EVP_PKEY(a, selection, "DER",
+                                               *output_structures,
+                                               libctx, propq);
+        if (ctx == NULL)
+            return -1;
+        if (OSSL_ENCODER_to_data(ctx, pp, &len))
+            ret = (int)len;
+        OSSL_ENCODER_CTX_free(ctx);
+        ctx = NULL;
+    }
+
+    if (ret == -1)
+        ERR_raise(ERR_LIB_ASN1, ASN1_R_UNSUPPORTED_TYPE);
+    return ret;
+}
+
+int i2d_KeyParams_ex(OSSL_LIB_CTX *libctx, const EVP_PKEY *a, const char *propq,
+                     unsigned char **pp)
+{
+    if (evp_pkey_is_provided(a)) {
+        const char *output_structures[] = { "type-specific", NULL };
+
+        return i2d_provided(libctx, a, propq, EVP_PKEY_KEY_PARAMETERS,
+                            output_structures, pp);
+    }
+    if (a->ameth != NULL && a->ameth->param_encode != NULL)
+        return a->ameth->param_encode(a, pp);
+    ERR_raise(ERR_LIB_ASN1, ASN1_R_UNSUPPORTED_TYPE);
+    return -1;
+}
+
+int i2d_KeyParams(const EVP_PKEY *a, unsigned char **pp)
+{
+    return i2d_KeyParams_ex(NULL, a, NULL, pp);
+}
+
+int i2d_KeyParams_bio(BIO *bp, const EVP_PKEY *pkey)
+{
+    return ASN1_i2d_bio_of(EVP_PKEY, i2d_KeyParams, bp, pkey);
+}
+
+int i2d_PrivateKey_ex(OSSL_LIB_CTX *libctx,
+                      const EVP_PKEY *a, const char *propq,
+                      unsigned char **pp)
+{
+    if (evp_pkey_is_provided(a)) {
+        const char *output_structures[] = { "type-specific", "pkcs8", NULL };
+
+        return i2d_provided(libctx, a, propq, EVP_PKEY_KEYPAIR,
+                            output_structures, pp);
+    }
+    if (a->ameth && a->ameth->old_priv_encode) {
```

Comment:
The legacy codepath has been block copied I assume..
It has a few coding standard issues (single line {}, != NULL, missing blanklines)

Reply:
Yes, they were copied verbatim

Reply:
Fixed